simulation/sub8_gazebo/sub8_gazebo_tools/bag_manager.py

```diff
@@ -0,0 +1,148 @@
+import rospy
+import rosbag
+import time
+import txros
+import os
+import yaml
+
+
+class BagManager(object):
+    def __init__(self, nh, diag_dir):
+        self.nh = nh
+        self.diag_dir = diag_dir
+
+        self.make_dict()
+        self.buffer_array_size = int(self.pre_cache_time / self.time_step)
+
+        self.cache_index = 0
+        self.cache = [None] * self.buffer_array_size
+        self.dumping = False
+
+    @txros.util.cancellableInlineCallbacks
+    def start_caching(self):
+        '''
+        Caches bags using big dicts.
+        {topic_name_1: subscriber_1, topic_name_2: subscriber_2, ...}
+        Every 'time_step' seconds it will save all of the subscribers' most recent message. This will continue and will fill
+            a cache array that will store 'pre_cache_time' seconds worth of messages before erasing the oldest messages.
+        '''
+
+        # Used to avoid adding copies of the same message.
+        last_timestamps = dict(self.cache_dict)
+        while True:
+            yield self.nh.sleep(self.time_step)
+
+            if self.dumping:
+                yield self.nh.sleep(1)
+
+            # Add to cache
+            msgs = []
+            for key in self.cache_dict:
+                msg = self.cache_dict[key].get_last_message()
+                msg_time = yield self.cache_dict[key].get_last_message_time()
+
+                if msg is not None:
+                    if (self.nh.get_time() > rospy.Time.from_sec(30)) and (msg_time < self.nh.get_time() - rospy.Duration(30)):
+                        # This fixes a negative time exception if we are within the first 30 seconds of time.
+                        continue
+
+                    if msg_time == last_timestamps[key]:
+                        continue
+                    last_timestamps[key] = msg_time
+
+                    msgs.append([key, (yield msg), (yield msg_time)])
+                else:
+                    print "There's a problem recording {0}".format(key)
+            self.write_to_cache(msgs)
+
+        self.dump()
+
+    @txros.util.cancellableInlineCallbacks
+    def make_dict(self):
+        '''
+        Generate caching dictionary from a yaml file.
+        '''
+        with open(self.diag_dir + 'messages_to_bag.yaml', 'r') as f:
+            messages_to_bag = yaml.load(f)
+
+        # Set bagging parameters
+        self.time_step = messages_to_bag['PARAMS']['time_step']
+        self.pre_cache_time = messages_to_bag['PARAMS']['pre_cache_time']
+        self.post_cache_time = messages_to_bag['PARAMS']['post_cache_time']
+
+        self.cache_dict = {}
+        msgs = messages_to_bag['MESSAGES']
+        for msg in msgs.values():
+            # Get message information
+            msg_topic = msg['message_topic']
+            msg_type = msg['message_type']
+            msg_name = msg['message_name']
+
+            # Import the message
+            exec("from {0}.msg import {1}".format(msg_type, msg_name))
```

Comment:
Nah, if you want to do it in the future you can use __import__, which is done in the __init__.py of gazebo_tests